salt/roster/terraform.py

```diff
@@ -0,0 +1,237 @@
+        base:
+          - srv/salt
+      pillar_roots:
+        base:
+          - srv/pillar
+      roster: terraform
+
+We setup a terraform file that creates N virtual machines (libvirt_domain resources) setup
+with cloud-init and the salt-ssh ssh key. In this case we set N to 2.
+
+.. code-block:: hcl
+
+    provider "libvirt" {
+      uri = "qemu:///system"
+    }
+
+    resource "libvirt_cloudinit" "init" {
+      name = "test-init.iso"
+      ssh_authorized_key = "${file("etc/salt/pki/master/ssh/salt-ssh.rsa.pub")}"
+    }
+
+    resource "libvirt_volume" "opensuse_leap" {
+      name = "opensuse_leap"
+      source = "http://r.opensu.se/Cloud:Images:Leap_42.3/images/openSUSE-Leap-42.3-OpenStack.x86_64.qcow2"
+    }
+
+    resource "libvirt_volume" "volume" {
+      name = "volume-${count.index}"
+      base_volume_id = "${libvirt_volume.opensuse_leap.id}"
+      count = 2
+    }
+
+    resource "libvirt_domain" "domain" {
+      name = "domain-${count.index}"
+      memory = 1024
+      disk {
+           volume_id = "${element(libvirt_volume.volume.*.id, count.index)}"
+      }
+
+      network_interface {
+        network_name = "default"
+        hostname = "minion${count.index}"
+        wait_for_lease = 1
+      }
+      cloudinit = "${libvirt_cloudinit.init.id}"
+      count = 2
+    }
+
+.. code-block:: bash
+
+    $ terraform apply
+    libvirt_cloudinit.init: Creating...
+    ...
+    libvirt_volume.opensuse_leap: Creating...
+    ...
+    libvirt_domain.domain.1: Creating...
+    libvirt_domain.domain.0: Creating...
+    ...
+    libvirt_domain.domain.0: Creation complete
+    ...
+    libvirt_domain.domain.1: Creation complete
+
+    Apply complete! Resources: 5 added, 0 changed, 0 destroyed.
+
+    The state of your infrastructure has been saved to the path
+    below. This state is required to modify and destroy your
+    infrastructure, so keep it safe. To inspect the complete state
+    use the `terraform show` command.
+
+    State path: terraform.tfstate
+
+You can start now using salt-ssh without any configuration.
+
+.. code-block:: bash
+
+   $ salt-ssh '*' test.ping
+   domain-0:
+       True
+   domain-1:
+       True
+
+'''
+from __future__ import absolute_import
+import logging
+import json
+import os.path
+import salt.utils
+from salt.roster.flat import RosterMatcher
+# pylint: disable=import-error
+from salt.ext import six
+# pylint: disable=redefined-builtin
+from salt.ext.six.moves import range
+# pylint: enable=import-error,redefined-builtin
+
+log = logging.getLogger(__name__)
+
+
+def _handle_libvirt_resource(resource, ret):
+    '''
+    Handles libvirt resources.
+    See https://github.com/dmacvicar/terraform-provider-libvirt
+    '''
+    primary = resource['primary']
+    attrs = primary['attributes']
+    name = attrs['name']
+    num_ifaces = int(attrs['network_interface.#'])
+    for if_k in range(num_ifaces):
+        num_addrs = int(attrs['network_interface.{}.addresses.#'.format(if_k)])
```

Comment:
A nitpick: in Salt we don't do empty braces for formatting, but at least one index (or name). I remember Pylint was also crying about this at least times ago.